Support php 8.0

axxapy opened the pull request:
Overview
----------
Initially this PR started same as https://github.com/php-amqplib/php-amqplib/pull/854 and https://github.com/php-amqplib/php-amqplib/pull/855
But during testing it I had to make certain changes to make tests run on php 8.0.

Changes
----------
 - Followed PHP 8.0 standards (no BC)
 - Added php 8.0 to github and travis flows
 - Made tests phpunit9.5-compatible (for php 8.0)
 - Adjusted one method (`AbstractIO::select`) to work properly under php 8.0 (problem was found by running tests under php 8.0)




Old description (for history, ignore it)
-------

> https://www.php.net/manual/en/migration80.deprecated.php
> 
> php 8.0 gives a lot of deprecation warnings:
> ```
> PHP Deprecated: Required parameter $source follows optional parameter $ticket in /srv/app/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Helper/Protocol/Protocol091.php on line 367
> PHP Deprecated: Required parameter $exchange follows optional parameter $ticket in /srv/app/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Helper/Protocol/Protocol091.php on line 432
> PHP Deprecated: Required parameter $exchange follows optional parameter $ticket in /srv/app/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Helper/Protocol/Protocol091.php on line 516
> PHP Deprecated: Required parameter $exchange follows optional parameter $reply_text in /srv/app/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Helper/Protocol/Protocol091.php on line 644
> PHP Deprecated: Required parameter $routing_key follows optional parameter $reply_text in /srv/app/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Helper/Protocol/Protocol091.php on line 644
> PHP Deprecated: Required parameter $exchange follows optional parameter $redelivered in /srv/app/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Helper/Protocol/Protocol091.php on line 662
> PHP Deprecated: Required parameter $routing_key follows optional parameter $redelivered in /srv/app/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Helper/Protocol/Protocol091.php on line 662
> ```
> 
> this PR fixes them

axxapy commented:
found out that even though it is done slightly different (mostly, remove default values instead of adding them), it is effectively a duplicate of https://github.com/php-amqplib/php-amqplib/pull/855

Please feel free to close it if you feel this way

Thank you guys

axxapy commented:
found out that even though it is done slightly different (mostly, remove default values instead of adding them), it is effectively a duplicate of https://github.com/php-amqplib/php-amqplib/pull/855

Please feel free to close it if you feel this way

Thank you guys

ramunasd commented:
This PR introduces a lot of breaking changes. We cannot accept it for next patch or minor version.

We will introduce PHP8.0 support only in next major version 3.0, not sooner. But at first, we will drop support for PHP <7.0. Also, everything related to connection constructors and parameters will be refactored. Specific code style and typehints will be must-have.

I can surely say that we won't accept any fast and not well tested BC.

If You @axxapy really interested in making this library better, then let's do it in small and clear steps. Do You agree?

axxapy commented:
Thank you for your reply. 

---
>This PR introduces a lot of breaking changes. 

Honestly speaking, I fail to see any breaking changes here:
 - tests will work as they used to be
 - all default params "in the middle" were not actually default ever. They served merely as a "hint" for developers but they never served as an actual default parameters: developers still had to explicitly set them due to the fact that following parameters were still required.
 - AbtsractConnection constructor already emitted InvalidArgumentException - already existing code which handles it should not change a bit.

---
>But at first, we will drop support for PHP <7.0

it would be better solution, I totally agree. The only problem I see here is that even if you get rid of php <7.0, you still won't support phpunit 9.3 (minimum version which supports php 8.0). phpunit 9.3 requires php 7.3 at least. Which means if you want to support php <7.3 => you'll still have to use same hacky workaround for tests I implemented in this PR.

---
(thinking out-loud):
The problem I have here is that php 8.0 exists "now", not "next year". A lot of common libraries (phpunit, for instance) have been preparing for php 8.0 _before_ it was released (since RC), not _after_.
If I were to maintain this repo, I'd say, in the current circumstances (php 8.0 has been already released) I can not afford to hold it's support till "next major release". **But I'm not a maintainer, I totally respect your, guys, vision.**

---
> If You @axxapy really interested in making this library better, then let's do it in small and clear steps. Do You agree?

If you want me to extract some part(s) of logic I introduced in this PR to a separate, smaller PR(s), please point on specific parts => I'll gladly do that

Yozhef commented:
I fully support @axxapy , I think it would be a good idea to test what is accepted now - and in version 3.0 to get rid of support <7.0 and so on - but that's just my opinion. Just infect a lot of projects to try to keep up with the news and everyone wants to work with PHP 8.0 =)

Baldinof commented:
Hi and happy new year :)

I do not see BC breaks either, the removed default arguments values were not usable and not part of the API. As @axxapy said, they can only serve as hints for the users. If it is still needed, these advice can be added to the phpdoc.

Baldinof commented on PhpAmqpLib/Wire/IO/AbstractIO.php:
```diff
@@ -82,11 +82,21 @@ public function select($sec, $usec)
-        } catch (\ErrorException $e) {
-            throw new AMQPIOWaitException($e->getMessage(), $e->getCode(), $e);
+
+        if (class_exists(\ValueError::class)) { // PHP 8.0+
```
You can use `PHP_VERSION_ID >= 80000` to avoid triggering the autoloader.

Baldinof commented on PhpAmqpLib/Connection/AbstractConnection.php:
```diff
@@ -180,11 +180,15 @@ public function __construct(
         $login_response = null,
         $locale = 'en_US',
-        AbstractIO $io,
+        AbstractIO $io = null,
```
Why is this needed?

dakujem replied:
see https://github.com/php-amqplib/php-amqplib/issues/846#issuecomment-752987677 for example, there are at least 3 other duplicate reports of the same issue.

Baldinof replied:
Sorry, I misexplained my thought.

This argument is already required (and not nullable) so all previous arguments must be provided.

It seems that removing the default value for all previous arguments could avoid the `is_null` check without breaking compatibility.

axxapy replied:
Basically I agree with you. It is not needed.
The single reason for this change is to make this PR minor-version-mergable.

dakujem commented:
For the new major release, I suggest dropping support for all EOL [PHP versions](https://www.php.net/supported-versions.php).

It would really be good if the current version supported PHP 8 as soon as possible, given this PR here clearly does not introduce a BC break, I see no reason not to accept and release it as a new MINOR version.

ramunasd commented:
Thanks for Your work. I will use it later on. For now I cannot merge this and the main reason is that this PR looks more like fast workaround rather than sophisticated solution. Some things can be changed now, but not everything from Your PR.

As I said, we won't introduce PHP8.0 support in 2.x series.

In my opinion, support for PHP8.0 is not just syntactical corectness. We must ensure that TCP connections and packet processing works as well. The things You didn't test, right?

I'm closing for now. Thanks for Your work again.

andrew-demb commented:
@ramunasd if the package doesn't support PHP 8, then composer.json should be updated to forbid package using with PHP 8

Now it is allowed - https://github.com/php-amqplib/php-amqplib/blob/master/composer.json#L32

ramunasd commented:
@andrew-demb sure, I will fix that.

axxapy commented:
> We must ensure that TCP connections and packet processing works as well. The things You didn't test, right?

I did actually, Can not be sure about all of small quirks and features but I have currently active project working on php8 and processing ~1000 rps. It is covered with functional tests => all seems to work fine.

ramunasd commented:
Thanks @axxapy , please let me know if You find any issues.

voyula commented:
Is there any update?

Baldinof commented:
Hi @lukebakken,

Thanks for re-opening this :) 

Does it mean we can expect a merge and a release of 3.0 soon?

lukebakken commented:
By the end of this week I hope. Please remember that the people who work on this project do so in their free time.

@axxapy it would be great to get the conflicts resolved in your branch. Thank you.

Baldinof commented:
Of course, it was just to have an estimate, take your time.

Thanks a lot for the efforts and the time you spend on maintaining this package!

ramunasd commented:
@lukebakken just saying that we still no ready with version 2.12.2. Are You planning to release something else meanwhile?

axxapy commented:
@lukebakken I have solved conflicts.
I had to merge master twice as between first and second merge another PR was merged to master.
I can rebase on master to avoid extra merge commits. If want me to do so, just let me know.
Thanks.

EtienneBruines commented:
Can someone manually restart https://travis-ci.org/github/php-amqplib/php-amqplib/jobs/758121133 ? It seemed to fail due to request limits. 

Alternatively, @axxapy -- the job/version/matrix definition of `.github/workflows/phpunit.yml` has changed, conflicting with this PR. Resolving this will probably start the tests again anyways (thus not requiring someone to manually restart the job).

ramunasd commented:
@axxapy please squash and rebase commits

amodliszewski commented:
Will it be merged anytime soon?

lukebakken commented:
I plan on releasing version 2.12.2 today. Then review and merge the PHP8 pull requests, then release 3.0.0. Most likely next week. Please be patient.

ramunasd commented on PhpAmqpLib/Wire/IO/AbstractIO.php:
```diff
@@ -82,11 +82,22 @@ public function select($sec, $usec)
+            try {
+                $result = $this->do_select($sec, $usec);
+                $this->cleanup_error_handler();
+            } catch (\Throwable $e) {
```
Why do we need to have separate try catch blocks? BTW, we are dropping PHP5.6 suuport, so let's one \Twroable block with additional check.

axxapy replied:
It is not for php 5.6. It is for php 8:
php 8 throws [ValueError](https://www.php.net/manual/en/class.valueerror.php) in this particular place. 
So correct `catch` block would be `catch (\ErrorException|\ValueError $ex)` (as it is said in the comments).
But such syntax is not supported by php 7.0 which is still supported. When 7.1 become minimum supported version this code could be removed.

masterjus replied:
https://www.php.net/supported-versions.php

PHP v7.3 is the minimal supported by the php community version. Is it necessary to support versions of PHP that would never get security updates?

ramunasd replied:
Why it's ValueError, where does this exception gets thrown? Are You sure we can replace ValueError with AMQPIOWaitException?

ramunasd replied:
@masterjus because there is still many with old PHP versions in production, despite it's not supported any more. Not everyone can upgrade easily.

axxapy replied:
@ramunasd I have reverted this commit => you can see the failure: https://github.com/php-amqplib/php-amqplib/pull/858/checks?check_run_id=1904199648

ramunasd replied:
Thanks @axxapy, this is a new issue, specific to PHP8. I will resolve it seperatelly as #883

axxapy replied:
So, what should I do?
Let me remove my revert so tests would pass => You'll fix it in a separate PR properly?

ramunasd replied:
Please keep it as it was orginally, thanks.

axxapy replied:
kk, I have rebased, dropped my commit. 
reference, just for the case: db1f905

ramunasd commented on tests/Functional/Bug/Bug256Test.php:
```diff
@@ -29,7 +27,7 @@ class Bug256Test extends AbstractConnectionTest
     protected $channel2;
 
-    public function setUp()
+    protected function setUpCompat()
```
I would suggest to use more meaningful method name, let's `initialize`. `setUpCompat` sounds like suspicious workaround :)

axxapy replied:
I strongly disagree.
1. `setUp` and `rearDown` methods - are standard phpunit's methods => any php developer who did tests at least once knows them. `initialize` - is something not standard. Every new developer who would want to contribute to this repo would have to learn "local" namings.
2. `Compat` pattern is pretty common. For instance, google uses it for [android compatibility libraries](https://developer.android.com/topic/libraries/support-library/packages). Adding `Compat` word to the end of the method/class names is self-explaining that it has something to do with compatibility.

Instead I:
- added phpdocs to these two methods so developers would see explanation why these methods exist.
- renamed `TestCase` to `TestCaseCompat` to avoid confusion between default phpunit's `TestCase` and "compat" `TestCase`.

ramunasd replied:
even more "compats", nice :D this phpunit comaptibility layer is temporal, I will drop it later on.

ramunasd commented on tests/bootstrap.php:
```diff
@@ -2,3 +2,9 @@
 require_once __DIR__ . '/config.php';
+
+if (PHP_VERSION_ID < 70100) {
+    require __DIR__ . '/php_compat_5.php';
```
We are dropping support for PHP5.6
Please do not use that version number.

ramunasd commented on .travis.yml:
```diff
@@ -9,6 +9,7 @@ matrix:
     - php: 7.3
       env: COVERAGE=true
     - php: 7.4
+    - php: 8.0
```
appveyor config needs to be updated as well

axxapy commented:
[note] the revert above is temporal and will be removed. Reason: https://github.com/php-amqplib/php-amqplib/pull/858#discussion_r576247985

axxapy commented:
Rebased, solved conflicts

Yozhef commented:
@axxapy can you please fix the conflict in Travis file. Now all CI in PHPUnit GitHubActions.

axxapy commented:
@Yozhef I have solved conflicts. 
Will it be merged now?

lukebakken reviewed (changes requested):
There is a test failure that remains when run on PHP8.

axxapy commented:
@lukebakken Tests failure is there only because I was explicitly *asked* to keep it:
https://github.com/php-amqplib/php-amqplib/pull/858#discussion_r576301324

Let me add my fix back...

lukebakken commented:
Apologies, I didn't follow that discussion.

masterjus commented:
I want to rise up a question regarding backwards compatibility on php versions. You say that many user didn't upgrade their php versions yet. OK, if there will be the major version number change they'll not be in trouble as composer will not upgrade php-amqplib for them. As you're already mentioned v^2 is stable, so it's no problem to leave it as a dependency.

It's strange that developers are not ready to upgrade version of PHP but ready to upgrade a library. I'm sure the've never run `composer update` because many other authors are not so conservative.

It's easy relation if you want upgrade `php-amqplib` => upgrade php to a minimal supported version.

Why am I so hardly want old versions drop? Because it removes a lot of messy code and makes a code much more safe and stable.

ramunasd commented:
@masterjus let me remind You, that this PR is about PHP8 support.
Please use discussions or mailing list for questions like Yours.

https://github.com/php-amqplib/php-amqplib/discussions
https://groups.google.com/forum/#!forum/php-amqplib-devel

ramunasd commented:
@axxapy I think there should be no more conflicts with other PR
Please rebase and squash all commits down to one.

axxapy commented:
I have solved conflicts, like fifth or so time. 
I'm 1000% okay to solve conflicts, but it is ~fifth or so time I do that. May I kindly ask you to ask me to solve conflicts next time right before you're actually going to merge the PR - so I did not do useless work? Thank you

> squash all commits down to one

Until PR is merged I'd rather prefer having separated commits so I could easily add/remove things there.

When you will merge this PR, please use [Squash and merge](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-request-merges#squash-and-merge-your-pull-request-commits) button - it will merge whole PR as a single commit.

![image](https://user-images.githubusercontent.com/3088476/108751862-c16ced80-7510-11eb-9a4a-808df00b0ea4.png)

axxapy commented:
@ramunasd you can take over if you feel so. You're the boss.
I'm okay to do rebases if there is sense in them.
My request was merely to minimize number of useless rebases.

Organizm238 commented:
@ramunasd 
Hi!
Sorry for pushing, but is there any estimated date when this MR will be merged? Our whole team is looking forward to this merge :)

keywinf commented:
@Organizm238 I may be wrong, but I did a `composer update` yesterday, and all went fine with php8.

AratKruglik commented:
@Organizm238 absolutely agree with you

lukebakken commented:
@ramunasd once the tests pass I will get `3.0.0`. It should be by the end-of-day tomorrow. Thanks @axxapy for these changes.

axxapy commented:
^^ Rebased to fresh master, dropped db1f905 again. Now there are even more failures. Please let me know what is the feather expected actions: 
 - return db1f905
 - fix failing tests in the scope of this PR (I can do that)
 - fix failing tests in the scope of another PR (#883 seems to be resolved, create new?)

ramunasd commented:
As far as I see, those exceptions appearrs only in PHP8, I need time to understand what is the reason.
@axxapy commit db1f905 obfuscates exceptions, does not fixes them, and creates wrong imagination that everything is ok. This is unwelcome way of development.

axxapy commented:
My humble opinion: these *new* failures are caused by the fact that #883 was fixed against master while it should be fixed against this PR. 
As a result, when #883 was fixed, php8 tests did not run => more failures were introduced than fixed.

#883 by definition was fixing errors in php8. It was strange to fix issues for php8 and merging "fix" to master without running tests against php8.

Saying that, will wait for your decision.

ramunasd commented:
thanks @axxapy , if You rebased from master, then my fix is in your branch as well? And final code would be same?
I did testing on PHP8 locally.
Those exceptions warns about connection stability issues. There might be platform specific problems. It's good that we found them now, not after release.

axxapy commented:
@ramunasd yes, I rebased to master, changes from #888 are here. Basically, new failures are caused by it.
#888 fixed `StreamIOTest` (I assume you tested it locally against php8, not the whole testsuite?) but broken other tests as now `do_select` method throws new exception (`AMQPConnectionClosedException`) which is not expected by them. So all broken test cases need to be updated

ramunasd commented:
Thanks @axxapy

amodliszewski commented:
@ramunasd When we can expect 3.0.0 release?

lukebakken commented:
@axxapy thanks again for all of your work on this!

@mihaileu @masterjus @andrew-demb 

Please take the time to test `v3.0.0-rc1` in your environment and report back if issues are found:

https://groups.google.com/g/rabbitmq-users/c/ktkiQ9slrME

https://packagist.org/packages/php-amqplib/php-amqplib#3.0.0-rc1

mihaileu commented:
Our tests were ok, we released a new tag on the bundle https://packagist.org/packages/emag-tech-labs/rabbitmq-bundle#2.5.0
